satellite/referrals/service.go

```diff
@@ -3,9 +3,195 @@
+	ErrUsedEmail = errs.Class("email used error")
+)
+
+// Config is for referrals service.
 type Config struct {
 	ReferralManagerURL storj.NodeURL
 }
+
+// Service allows communicating with the Referral Manager
+//
+// architecture: Service
+type Service struct {
+	log          *zap.Logger
+	signer       signing.Signer
+	config       Config
+	dialer       rpc.Dialer
+	db           console.Users
+	passwordCost int
+}
+
+// NewService returns a service for handling referrals information.
+func NewService(log *zap.Logger, signer signing.Signer, config Config, dialer rpc.Dialer, db console.Users, passwordCost int) *Service {
+	return &Service{
+		log:          log,
+		signer:       signer,
+		config:       config,
+		dialer:       dialer,
+		db:           db,
+		passwordCost: passwordCost,
+	}
+}
+
+// GetTokens returns tokens based on user ID.
+func (service *Service) GetTokens(ctx context.Context, userID *uuid.UUID) (tokens []uuid.UUID, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if userID.IsZero() {
+		return nil, errs.New("invalid argument")
+	}
+
+	conn, err := service.referralManagerConn(ctx)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	defer func() {
+		err = conn.Close()
+	}()
+
+	client := conn.ReferralManagerClient()
+	response, err := client.GetTokens(ctx, &pb.GetTokensRequest{
+		OwnerUserId: userID[:],
+	})
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	tokensInBytes := response.GetTokenSecrets()
+	if len(tokensInBytes) == 0 {
+		return nil, errs.New("no available tokens")
+	}
+
+	tokens = make([]uuid.UUID, len(tokensInBytes))
+	for i := range tokensInBytes {
+		token, err := bytesToUUID(tokensInBytes[i])
+		if err != nil {
+			continue
+		}
+		tokens[i] = token
+	}
+
+	return tokens, nil
+}
+
+// CreateUser validates user's registration information and creates a new user.
+func (service *Service) CreateUser(ctx context.Context, user CreateUser) (_ *console.User, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if err := user.IsValid(); err != nil {
+		return nil, ErrValidation.Wrap(err)
+	}
+
+	if len(user.ReferralToken) == 0 {
+		return nil, errs.New("invalid argument")
+	}
+
+	_, err = service.db.GetByEmail(ctx, user.Email)
+	if err == nil {
+		return nil, ErrUsedEmail.New("")
+	}
+
+	userID, err := uuid.New()
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	err = service.redeemToken(ctx, userID, user.ReferralToken)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	hash, err := bcrypt.GenerateFromPassword([]byte(user.Password), service.passwordCost)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	newUser := &console.User{
+		ID:           *userID,
+		Email:        user.Email,
+		FullName:     user.FullName,
+		ShortName:    user.ShortName,
+		PasswordHash: hash,
+	}
+
+	u, err := service.db.Insert(ctx,
+		newUser,
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return u, nil
+}
+
+func (service *Service) redeemToken(ctx context.Context, userID *uuid.UUID, token string) error {
```

Comment:
why is this method unexported but the other ones are exported?

Reply:
because in order to redeem a token, we need user's uuid. At the moment, to redeem a token, we would call Register